<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN"
    "http://www.w3.org/TR/html4/strict.dtd">
<html lang=en>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<title>GLAM2 Manual</title>
<link type="text/css" rel="stylesheet" href="glam2.css">
    <link rel="icon" type="image/png" href="../images/memesuite_icon.png">
</head>
<body>

<img src="images/glam2_logo.png" alt="glam2 logo">

<h1>GLAM2 Manual</h1>

<p><strong>Summary:</strong> GLAM2 is a software package for finding
<em>motifs</em> in sequences, typically amino-acid or nucleotide
sequences. A motif is a re-occurring sequence pattern: typical examples
are the TATA box and the CAAX prenylation motif. The main innovation of
GLAM2 is that it allows insertions and deletions in motifs.</p>

<p><strong>Family:</strong> GLAM2 is a child of <a
	href="http://zlab.bu.edu/glam/">GLAM</a> and a sibling of <a
	href="ftp://ftp.ncbi.nih.gov/pub/spouge/papers/archive/AGLAM/">A-GLAM</a>.
If you are looking for nucleotide motifs without insertions or
deletions, A-GLAM is probably better. 'GLAM' used to stand for Gapless
Local Alignment of Multiple sequences, but it now stands for Gapped
Local Alignment of Motifs.</p>

<p<strong>Inventory:</strong> The package includes these programs:</p>

<ul>

	<li><strong>glam2</strong> - for discovering motifs shared by a
	set of sequences.</li>

	<li><strong>glam2scan</strong> - for finding matches, in a
	sequence database, to a motif discovered by glam2.</li>

	<li><strong>glam2format</strong> - for converting glam2 motifs to
	standard alignment formats.</li>

	<li><strong>glam2mask</strong> - for masking glam2 motifs out of
	sequences, so that weaker motifs can be found.</li>

	<li><strong>purge</strong> - for removing highly similar members
	of a set of sequences.</li>

</ul>

<p>These are <em>command line</em> programs, so you need a basic
familiarity with running things from your computer's command line in
order to use them. (In Mac OS X, you want Applications -> Utilities ->
Terminal.)</p>

<p><strong>About:</strong> GLAM2 was developed by <a
	href="http://www.cbrc.jp/~martin/">Martin C Frith</a>, working at the <a
	href="http://www.cbrc.jp/index.eng.html">Computational Biology
Research Center</a> in Tokyo, and <a
	href="http://research.imb.uq.edu.au/~tbailey/">Timothy L Bailey</a>,
working at the <a href="http://www.imb.uq.edu.au/">Institute for
Molecular Bioscience</a> in Brisbane. The source code and documentation are
hereby released into the public domain. If you use GLAM2, please cite: <cite>MC
Frith, NFW Saunders, B Kobe, TL Bailey (2008) Discovering sequence
motifs with arbitrary insertions and deletions, PLoS Computational
Biology, <b>4</b>(5), e1000071, 2008. </cite>.</p>

<p><strong>Contact:</strong> Questions and comments are welcome.
Email: glam2(at-mark)imb.uq.edu.au. We would especially like to hear
about any errors when installing or running the software (e.g.
segmentation faults, assertion failures) so that we can fix them. Please
send the exact text of the error message, and the exact inputs that
produce the error.</p>

<h2>Contents</h2>

<ul>

	<li><a href="glam2_install.html">Installation</a> - how to install
	GLAM2 on your computer.</li>

	<li><a href="glam2_tut.html">glam2 tutorial</a> - all you need to
	know for typical usage of glam2.</li>

	<li><a href="glam2_ref.html">glam2 reference</a> - all of glam2's
	options.</li>

	<li><a href="glam2scan_man.html">glam2scan manual</a></li>

	<li><a href="glam2format_man.html">glam2format manual</a></li>

	<li><a href="glam2mask_man.html">glam2mask manual</a></li>

	<li><a href="purge.html">purge manual</a></li>

	<li><a href="glam2_alphabet.html">Alphabets</a></li>

	<li><a href="glam2_dirichlet.html">Dirichlet mixtures</a></li>

	<li><a href="GLAM2_method.pdf">GLAM2 methods (PDF)</a> -
	mathematical description of how GLAM2 works.</li>

</ul>

</body>
</html>
