<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
                      "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"
>
<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <title>MCAST - Introduction</title>
    <link href="doc/meme-suite.css" rel="stylesheet" type="text/css" />
    <link rel="shortcut icon" type="images/x-icon" href="images/memesuite_icon.ico">
  </head>
  <body class="body">
    <table class="maintable">
      <tr>
        <td class="menubase">
          <div id="menu">
            <script src="meme-suite-menu.js" type="text/javascript"></script>
          </div>
        </td>
        <td class="maintablewidth">
          <div id="main">
            <script src="doc/meme-suite-logo.js" type="text/javascript"></script>
            <noscript>
              <h1>MEME Suite</h1>
              The MEME Suite web application requires the use of JavaScript
              <br />
              Javascript doesn't seem to be available on your browser.
            </noscript>
	          <img src="doc/images/mcast-logo.png" alt="MCAST logo">
            <h3>Introduction</h3>
            <h4><code>MCAST</code> searches a sequence database for
            statistically significant clusters of non-overlapping "hits" to the
            motifs in a query.</h4>
            <p>A "hit" is a sequence position that is sufficiently similar to a
            motif in the query. To be a hit, the <i>p</i>-value of the motif
            alignment score must be less than the significance threshold,
            <i>pthresh</i> (see optiontal input <i>p</i>-value threshold, below). 
            The alignment of the motif
            and the sequence position is done <b>without</b> gaps. To compute
            the <i>p-</i>value of a motif alignment score, <code>MCAST</code>
            assumes that the sequences in the database were generated by a
            0-order Markov process. 
            <code>MCAST</code> searches for hits on both the sequences given in
            the database, and their reverse complements.</p>
            <p>A cluster of non-overlapping hits is called a "match". 
            The user can specify the maximum allowed distance between the 
            hits in a match (see optional input maximum motif gap below).
            Two hits separated by more than the maximum allowed gap will be
            reported in separate matches.</p>
            <p><code>MCAST</code> searches for all of the matches between the
            query and the sequences in the database. Each match is assigned an
            <i>E</i>-value, and matches that score below an <i>E</i>-value
            threshold are printed in order of increasing <i>E</i>-value (see
            optional input <i>E</i>-value threshold below).</p>
            <p>The <i>p-</i>value of a hit is converted to a "p-score" in order
            to compute the total score of the match it participates in. 
            The p-score for a hit with <i>p-</i>value <i>p</i> is</p>
            <div style="margin-left: 2em"><p><i>S =
            -log<sub>2</sub>(p/pthresh),</i></p></div>
            <p>
            where the significance threshold <i>pthresh</i> may be specified by
            the user (see optional input motif <i>p</i>-threshold below. 
            The total score of a match is the sum of the p-scores of
            the hits making up the match. <code>MCAST</code> finds the matches
            with the maximum match scores.
            </p>
            <p>In order for <i>E</i>-values to be computed by
            <code>MCAST</code>, at least 100 matches must be found. If there
            are too few sequences in the database, or if certain other options
            are made to stringent (see Options, below), too few matches may
            exist for <i>E</i>-values to be computed. In this case, the results
            are sorted by match score, the <i>E</i>-value column is set to
            "NaN" and all matches are printed.</p>
            <p>
            A full description of the algorithm is found in:
            </p>
            <blockquote>
            <p>
            <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/19/suppl_2/ii16">
            Bailey and Noble. "Searching for statistically significant regulatory
            modules." <i>Bioinformatics (Proceedings of the European Conference on
            Computational Biology)</i>. 19(Suppl. 2):ii16-ii25, 2003.</a>
            </p>
            </blockquote>
            </p>
            <h4>Required MCAST Inputs</h4>
            <p/>
              Three inputs must be provided on the MCAST web page:
              <ol>
              <li>
              An e-mail address where the notification of job completion can be sent.
              You specify the e-mail address in the two text boxes labeled "e-mail address". 
              The e-mail address must be entered twice to reduce the amount of
              undeliverable mail caused by typographic errors.
              </li>
              <li>
              A <a href="./doc/meme-format.html">MEME output file</a>, 
              containing the descriptions of one or more motifs.
              You can select a file to be uploaded from your computer by clicking on the 
              "Browse ..." button under the "Your motif file" label.
              </li>
              <li>
              A sequence database to be searched. You can choose a sequence file to be
              uploaded from your computer bu clicking on the "Browse ..." button under the "Your
              FASTA sequence file" label.
              Alternatively, you can select one of the supported databases maintained on
              the MEME Suite web site:
              first select the category of the
              sequence database from the "Category" drop-down list, then choose one of
              the supported databases listed in the "Database" drop-down list.
              </li>
              </ol>
            <h4>Optional MCAST inputs</h4>
            <p/>
            The MCAST web page accepts four optional inputs:
            <ol>
            <li>A threshold p-value. Motif occurrences with p-values below the threshold
            will not be considered in scoring matches (defaults to <code>5e-4</code>).</li>
            <li>A motif gap maximum. The maximum allowed distance between adjacent motif 
            hits in a match (defaults to 50).</li>
            <li>The E-value threshold. Matches whose E-value is less then the threshold
            will not be reported (defaults to 10).</li>
            <li>The pseudocount weight. A pseudocount is added to each count in the 
            motif matrix. The pseudocount is determined by multiplying the background
            frequency by this weight (defaults to 4).</li>
            </ol>
            <p/>
            The contents of the required and optional fields can be cleared by clicking on
            the "Clear Input" button.
            <h4>Submitting an MCAST job</h4>
            <p/>
            After filling in the required and optional inputs on the MCAST web page,
            the job can be submitted for processing by clicking on the "Start Search"
            button.
            <h4>MCAST output</h4>
            <p/>
            The result of submitting your job will be a summary page that briefly describes 
            your job's input, or reports any critical errors.
            If there are no errors, then the page will also contain a link to your job's results.
            The results page will report the full command that was run
            and will include links 
            to the MCAST output, or to any error messages that were generated.
            <p/>
            The MCAST output will be an HTML file 
            (<a href ="doc/examples/mcast_example_output_files/mcast.html">sample</a>)
            containing a table of clusters of the input motifs, 
            sorted by E-value, followed by alignment diagrams showing the motif
            hits in each cluster.
            <script src="template-footer.js" type="text/javascript"></script>
          </div>
        </td>
      </tr>
    </table>
  </body>
</html>
